registry.go

```diff
@@ -108,3 +112,9 @@ type SignatureService interface {
 	// Put stores the signature for the provided digest.
 	Put(dgst digest.Digest, signatures ...[]byte) error
 }
+
+// CatalogService provides a way of retrieving the names of each of the repositories
+type CatalogService interface {
+	// Get retrieves repository names from the registry.
+	Get(n int, q string) (p []string, moreEntries bool, err error)
```

Comment:
Rather than `moreEntries`, return an integer with the number of values populated in `p`. This can avoid a ton of allocations, since `p` can be reused:

```
Repositories(p []string, q string) (n int, err error)
```

So, up to len(p) will be filled, with `n` denoting how many were actually filled. Signal completion with the err value `io.EOF` or something explicit.